### DISCUSSION TITLE
Support for Execution Providers - QNN

### DISCUSSION THREAD

**Discussion by adhurwit** [microsoft/onnxruntime-genai#660] [Support for Targets (OS / EPs / Hardware)]
In order to make use of Qualcomm's NPU, we need to use the ONNX EP for QNN. There does not appear to be a way to specify that right now. Please add support for Qualcomm NPUs

**Comment by natke** [comment]
Hi @adhurwit, we are working on this! Stay tuned.

**Reply by adhurwit** [reply]
how is this coming along? is there a target date?

**Comment by natke** [comment]
Moving into Discussion

**Reply by Davidqian123** [reply]
Any update for QNN Execution Providers? Appreciate a lot!

**Comment by axcpy08** [comment]
Hi there. Do you have ETA for QNN EP? Thanks!

**Comment by hmartinez82** [comment]
Is there anything I can do to help? I can't even find instructions on how to build genai with QNN support enabled. I already have onnxruntime built locally for QNN.

**Comment by scsonic** [comment]
there were a branch named phi3.5_qnn_test / android
it's possible to share model builder with qnn?

**Comment by jasonw247** [comment]
Is there any update on this? Are there plans to support context binaries/split models for HTP to get around HTP rpc memory size limitations?

**Reply by baijumeswani** [reply]
Please take a look at these resources:

- https://onnxruntime.ai/docs/genai/tutorials/snapdragon.html
- https://onnxruntime.ai/docs/genai/howto/build-models-for-snapdragon.html

**Reply by axcpy08** [reply]
Thanks for providing the resources. I followed the instruction and faced the RuntimeError as below picture. Any suggestions for solving it?

![Screenshot 2025-03-18 163443](https://github.com/user-attachments/assets/141d4104-3de8-4296-a3a5-e64171f35965)

**Reply by jasonw247** [reply]
Thanks! Are LoRa adapters supported with the QNN execution provider?

I see [this](https://github.com/microsoft/Olive/issues/1544#issuecomment-2605509758) issue  which notes that `we only verified the example with CPU and CUDA EP`. Is this the current state?

I also see: https://onnxruntime.ai/docs/execution-providers/QNN-ExecutionProvider.html#lorav2-support. Is there more documentation around this? Does this support adapter loading/unloading/switching?

Is that compatible with: https://onnxruntime.ai/blogs/multilora?

**Reply by baijumeswani** [reply]
> I see https://github.com/microsoft/Olive/issues/1544#issuecomment-2605509758 issue which notes that we only verified the example with CPU and CUDA EP. Is this the current state?

This is still accurate. We have not done validation with DML or QNN execution providers yet. As such CPU and Cuda are the primary execution providers where lora support can be expected to work without hiccups.
